### PR TITLE
array subjob once rejected by runjob hook will never run

### PR DESCRIPTION
**Pull request by Bhagat-Rajput**
#### Bug/feature Description
* array subjob once rejected by runjob hook will never run
* *[PP-XXXX](https://pbspro.atlassian.net/browse/PP-XXXX)*

#### Affected Platform(s)
* All

#### Cause / Analysis / Design
* Runjob hook was rejecting the array subjobs after the job structure creation(**create_subjob()**), so in this case subjob structure was saved.

#### Solution Description
* Need to free the subjob structure also, if subjob has been rejected by runjob hook.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/server/job_func.c [hunk 1]
```diff
@@ -775,7 +775,7 @@ job_purge(job *pjob)
 		(pjob->ji_qs.ji_substate != JOB_SUBSTATE_TRANSICM)) {
 		if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_SubJob) {
 			svr_dequejob(pjob);
-			if (pjob->ji_qs.ji_substate == JOB_SUBSTATE_RERUN3)
+			if ((pjob->ji_qs.ji_substate == JOB_SUBSTATE_RERUN3) || (pjob->ji_qs.ji_substate == JOB_SUBSTATE_QUEUED))
```
Comment (arungrover) [inline src/server/job_func.c]: why is this check for queued substate needed?

File: test/tests/functional/pbs_runjob_hook.py [hunk 2]
```diff
@@ -38,19 +38,23 @@
...
 j = e.job
 pbs.logmsg(pbs.LOG_DEBUG, "job_id=%s" % j.id)
 pbs.logmsg(pbs.LOG_DEBUG, "sub_job_array_index=%s"
            % j.array_index)
 e.accept()
+"""
+
+    reject_hook_script = """
+import pbs
+pbs.event().reject("Rejecting job by runjob hook event")
```
Comment (arungrover) [inline test/tests/functional/pbs_runjob_hook.py]: Can you please change the message to "runjob hook rejected the job"

File: test/tests/functional/pbs_runjob_hook.py [hunk 3]
```diff
@@ -78,9 +83,33 @@ def test_normal_job_index(self):
...
+        jobs should run after runjob hook enabled set to false.
+        """
+        hook_name = "runjob_hook"
+        attrs = {'event': "runjob"}
+        rv = self.server.create_import_hook(hook_name, attrs,
+                                            self.reject_hook_script,
+                                            overwrite=True)
+        self.assertTrue(rv)
+        j1 = Job(TEST_USER)
+        lower = 1
```
Comment (arungrover) [inline test/tests/functional/pbs_runjob_hook.py]: why do you need these variables? just give '1-3' in the dictionary

File: test/tests/functional/pbs_runjob_hook.py [hunk 4]
```diff
@@ -78,9 +83,33 @@ def test_normal_job_index(self):
...
+        upper = 3
+        j1.set_attributes({ATTR_J: '%d-%d' % (lower, upper)})
+        jid = self.server.submit(j1)
+        msg = "Not Running: PBS Error: Rejecting job by runjob hook event"
+        self.server.expect(JOB, {'job_state': 'Q', 'comment': msg}, id=jid)
+        a = {'enabled': 'false'}
+        self.server.manager(MGR_CMD_SET, HOOK, a, id=hook_name,
+                            expect=True, sudo=True)
+        self.server.expect(JOB, {'job_state': 'B'}, id=jid)
```
Comment (arungrover) [inline test/tests/functional/pbs_runjob_hook.py]: don't you need to trigger a scheduling cycle here?

Comment (Bhagat-Rajput) [reply test/tests/functional/pbs_runjob_hook.py]: Thanks @arungrover. Yes, i forgot to trigger the scheduling cycle but expect function was triggering itself that's why testcase got passed.
I'll add the needful change.

File: src/server/job_func.c (hunk 1 above)
Comment (Bhagat-Rajput) [reply src/server/job_func.c]: @arungrover it's needed becasue before calling call_to_process_hooks, create_subjob() changed the state(JOB_STATE_QUEUED to JOB_SUBSTATE_QUEUED) but in job_purge() we were checking for JOB_SUBSTATE_RERUN3 state only.

**Comment by Bhagat-Rajput** [comment]
Rebased and squashed, ready to merge.

**Comment by mike0042** [comment]
I don't see any test logs, but the box is checked.

**Comment by Bhagat-Rajput** [comment]
Sorry @mike0042 i forgot to attach the file, have attached the PTL logs file.
[ptl_runjob_hook_logs.txt](https://github.com/PBSPro/pbspro/files/1842581/ptl_runjob_hook_logs.txt)

**Comment by arungrover** [comment]
@Bhagat-Rajput Please rebase again

**Comment by Bhagat-Rajput** [comment]
@arungrover  Rebased.

**Pull request merged**